File: tools/mujs/lib/ast/expressions.ts
// Copyright 2016 Marapongo, Inc. All rights reserved.

import {Identifier, Node} from "./nodes";
import * as statements from "./statements";

import * as symbols from "../symbols";

export interface Expression extends Node {}

/** Literals **/

export interface Literal extends Expression {
    raw?: string; // the raw literal, for round tripping purposes.
}

// A `null` literal.
export interface NullLiteral extends Literal {
    kind: NullLiteralKind;
}
export const nullLiteralKind = "NullLiteral";
export type  NullLiteralKind = "NullLiteral";

// A `bool`-typed literal (`true` or `false`).
export interface BoolLiteral extends Literal {
    kind:  BoolLiteralKind;
    value: boolean;
}
export const boolLiteralKind = "BoolLiteral";
export type  BoolLiteralKind = "BoolLiteral";

// A `number`-typed literal (floating point IEEE 754).
export interface NumberLiteral extends Literal {
    kind:  NumberLiteralKind;
    value: number;
}
export const numberLiteralKind = "NumberLiteral";
export type  NumberLiteralKind = "NumberLiteral";

// A `string`-typed literal.
export interface StringLiteral extends Literal {
    kind:  StringLiteralKind;
    value: string;
}
export const stringLiteralKind = "StringLiteral";
export type  StringLiteralKind = "StringLiteral";

// A array literal (`new` and/or initialization).
export interface ArrayLiteral extends Literal {
    kind:      ArrayLiteralKind;
    type:      symbols.TypeToken; // the type of array to produce.
    size?:     Expression;        // an optiojnal expression for the array size.
    elements?: Expression[];      // an optional array of element expressions to store into the array.
}
export const arrayLiteralKind = "ArrayLiteral";
export type  ArrayLiteralKind = "ArrayLiteral";

// An object literal (`new` and/or initialization).
export interface ObjectLiteral extends Literal {
    kind:          ObjectLiteralKind;
    type:          symbols.TypeToken;          // the type of object to produce.
    initializers?: ObjectLiteralInitializer[]; // an optional array of property initializers.
    arguments?:    Expression[];               // an optional set of arguments for the constructor.
}
export const objectLiteralKind = "ObjectLiteral";
export type  ObjectLiteralKind = "ObjectLiteral";

// An object literal property initializer.
export interface ObjectLiteralInitializer extends Node {
    kind:     ObjectLiteralInitializerKind;
    property: symbols.VariableToken; // the property being initialized.
    value:    Expression;            // the expression value to store into the property.
}
export const objectLiteralInitializerKind = "ObjectLiteralInitializer";
export type  ObjectLiteralInitializerKind = "ObjectLiteralInitializer";

/** Loads **/

// TODO(joe): figure out how to load/store elements and maps.  Possibly just use intrinsic functions.

export interface LoadExpression extends Expression {
}

// Loads a location's address, producing a pointer that can be dereferenced.
export interface LoadLocationExpression extends LoadExpression {
    kind:    LoadLocationExpressionKind;
    object?: Expression; // the `this` object, in the case of class properties.
    name:    Identifier; // the name of the member to load.
}
export const loadLocationExpressionKind = "LoadLocationExpression";
export type  LoadLocationExpressionKind = "LoadLocationExpression";

// Dynamically loads either a variable or function, by name, from an object.
// TODO(joe): I'm unsure if we should permit assigning to functions by name; I think we'll need to for Python/Ruby/etc.
export interface LoadDynamicExpression extends LoadExpression {
    kind:   LoadDynamicExpressionKind;
    object: Expression; // the object to load a property from.
    name:   Expression; // the name of the property to load.
}
export const loadDynamicExpressionKind = "LoadDynamicExpression";
export type  LoadDynamicExpressionKind = "LoadDynamicExpression";

/** Functions **/

export interface CallExpression extends Expression {
    arguments?: Expression[]; // the list of arguments in sequential order.
}

// Allocates a new object and calls its constructor.
export interface NewExpression extends CallExpression {
    kind: NewExpressionKind;
    type: Identifier; // the object type to allocate.
}
export const newExpressionKind = "NewExpression";
export type  NewExpressionKind = "NewExpression";

// Invokes a function.
export interface InvokeFunctionExpression extends CallExpression {
    kind:     InvokeFunctionExpressionKind;
    function: Expression; // a function to invoke (of a func type).
}
export const invokeFunctionExpressionKind = "InvokeFunctionExpression";
export type  InvokeFunctionExpressionKind = "InvokeFunctionExpression";

// Creates a lambda, a sort of "anonymous" function.
export interface LambdaExpression extends Expression {
    kind:       LambdaExpressionKind;
    signature:  symbols.TypeToken;       // the func signature type.
    parameters: symbols.VariableToken[]; // the parameter variables.
    body:       statements.Block;        // the lambda's body block.
}
export const lambdaExpressionKind = "LambdaExpression";
export type  LambdaExpressionKind = "LambdaExpression";

/** Operators **/

// A unary operator expression.
export interface UnaryOperatorExpression extends Expression {
    kind:     UnaryOperatorExpressionKind;
    operator: UnaryOperator; // the operator type.
    operand:  Expression;    // the right hand side operand.
    postfix:  boolean;       // whether this is a postifx operator (only legal for UnaryPfixOperator).
}
export const unaryOperatorExpressionKind = "UnaryOperatorExpression";
export type  UnaryOperatorExpressionKind = "UnaryOperatorExpression";

// A unary prefix/postfix operator token.
export type UnaryPfixOperator = "++" | "--";

// A unary operator token.  Note that MuIL doesn't care about precedence.  The MetaMu compilers must present the
// expressions in the order in which they should be evaluated through an in-order AST tree walk.
export type UnaryOperator = UnaryPfixOperator |
                            "*" | "&"         | // dereference and addressof.
                            "+" | "-"         | // unary plus and minus
                            "!" | "~"         ; // logical NOT and bitwise NOT.

// A binary operator expression (assignment, logical, operator, or relational).
export interface BinaryOperatorExpression extends Expression {
    kind:     BinaryOperatorExpressionKind;
    left:     Expression;     // the left hand side.
    operator: BinaryOperator; // the operator.
    right:    Expression;     // the right hand side.
}
export const binaryOperatorExpressionKind = "BinaryOperatorExpression";
export type  BinaryOperatorExpressionKind = "BinaryOperatorExpression";

// All of the available arithmetic operators.
export type BinaryArithmeticOperator  = "+"   | "-"   | // addition and subtraction.
                                        "*"   | "/"   | // multiplication and division.
                                        "%"   | "**"  ; // remainder and exponentiation.

// All of the available assignment operators.
// TODO: figure out what to do with ECMAScript's >>>= operator.
export type BinaryAssignmentOperator  = "="           | // simple assignment.
                                        "+="  | "-="  | // assignment by sum and difference.
                                        "*="  | "/="  | // assignment by product and quotient.
                                        "%="  | "**=" | // assignment by remainder and exponentiation.
                                        "<<=" | ">>=" | // assignment by bitwise left and right shift.
                                        "&="  | "|="  | // assignment by bitwise AND and OR.
                                        "^="          ; // assignment by bitwise XOR.

// All of the available bitwise operators.
// TODO: figure out what to do with ECMAScript's >>> operator.
export type BinaryBitwiseOperator     = "<<"  | ">>"  | // bitwise left and right shift.
                                        "&"   | "|"   | // bitwise AND and OR (inclusive OR).
                                        "^"           ; // bitwise XOR (exclusive OR).

// All of the available conditional operators.
export type BinaryConditionalOperator = "&&"  | "||"  ; // logical AND and OR.

// All of the available relational operators.
// TODO: figure out what to do with ECMAScript's === and !=== operators.
export type BinaryRelationalOperator  = "<"   | "<="  | // relational operators less-than and less-than-or-equals.
                                        ">"   | ">="  | // relational operators greater-than and greater-than-or-equals.
                                        "=="  | "!="  ; // relational operators equals and not-equals.

// A binary operator token.  Note that MuIL doesn't care about precedence.  The MetaMu compilers must present the
// expressions in the order in which they should be evaluated through an in-order AST tree walk.
export type BinaryOperator = BinaryArithmeticOperator  |
                             BinaryAssignmentOperator  |
                             BinaryBitwiseOperator     |
                             BinaryConditionalOperator |
                             BinaryRelationalOperator  ;

/** Type Testing **/

// A cast expression; this handles both nominal and structural casts, and will throw an exception upon failure.
export interface CastExpression extends Expression {
    kind:       CastExpressionKind;
    expression: Expression;        // the source expression.
    type:       symbols.TypeToken; // the target type.
}
export const castExpressionKind = "CastExpression";
export type  CastExpressionKind = "CastExpression";

// An isinst expression checks an expression for compatibility with the given type token, evaluating to a boolean.
export interface IsInstExpression extends Expression {
    kind:       IsInstExpressionKind;
    expression: Expression;        // the source expression.
    type:       symbols.TypeToken; // the target type.
}
export const isInstExpressionKind = "IsInstExpression";
export type  IsInstExpressionKind = "IsInstExpression";

// An typeof instruction gets the type token -- just a string -- of a particular expression at runtime.
export interface TypeOfExpression extends Expression {
    kind:       TypeOfExpressionKind;
    expression: Expression;        // the source expression.
}
export const typeOfExpressionKind = "TypeOfExpression";
export type  TypeOfExpressionKind = "TypeOfExpression";

/** Miscellaneous **/

// A conditional expression.
export interface ConditionalExpression extends Expression {
    kind:       ConditionalExpressionKind;
    condition:  Expression; // a `bool` condition expression.
    consequent: Expression; // the expression to evaluate if `true`.
    alternate:  Expression; // the expression to evaluate if `false`.
}
export const conditionalExpressionKind = "ConditionalExpression";
export type  ConditionalExpressionKind = "ConditionalExpression";

// A sequence expression allows composition of multiple expressions into one.  It evaluates to the last one.
export interface SequenceExpression extends Expression {
    kind:        SequenceExpressionKind;
    expressions: Expression[];
}
export const sequenceExpressionKind = "SequenceExpression";
export type  SequenceExpressionKind = "SequenceExpression";

File: pkg/pack/ast/expressions.go
// Copyright 2016 Marapongo, Inc. All rights reserved.

package ast

import (
	"github.com/marapongo/mu/pkg/pack/symbols"
)

// Expression is an executable operation that usually produces a value.
type Expression interface {
	Node
	expression()
}

type expression struct {
	node
}

func (node *expression) expression() {}

/* Literals */

type Literal interface {
	GetRaw() *string // the raw literal, for round tripping purposes.
}

type literal struct {
	expression
	Raw *string `json:"raw,omitempty"`
}

func (node *literal) GetRaw() *string { return node.Raw }

// NullLiteral represents the usual `null` constant.
type NullLiteral struct {
	literal
}

// BoolLiteral represents the usual Boolean literal constant (`true` or `false`).
type BoolLiteral struct {
	literal
	Value bool `json:"value"`
}

// NumberLiteral represents a floating point IEEE 754 literal value.
type NumberLiteral struct {
	literal
	Value float64 `json:"value"`
}

// StringLiteral represents a UTF8-encoded string literal.
type StringLiteral struct {
	literal
	Value string `json:"value"`
}

// ArrayLiteral evaluates to a newly allocated array, with optional initialized elements.
type ArrayLiteral struct {
	literal
	Type     *symbols.TypeToken `json:"type,omitempty"`     // the optional type of array being produced.
	Size     *Expression        `json:"size,omitempty"`     // an optional expression for the array size.
	Elements *[]Expression      `json:"elements,omitempty"` // an optional array of element initializers.
}

// ObjectLiteral evaluates to a new object, with optional property initializers for primary properties.
type ObjectLiteral struct {
	literal
	Type       *symbols.TypeToken       `json:"type,omitempty"`       // the optional type of object to produce.
	Properties *[]ObjectLiteralProperty `json:"properties,omitempty"` // an optional array of property initializers.
}

// ObjectLiteralProperty initializes a single object literal property.
type ObjectLiteralProperty struct {
	node
	Name  *Identifier `json:"name"`  // the property to initialize.
	Value Expression  `json:"value"` // the expression whose value to store into the property.
}

/* Loads */

type LoadExpression interface {
	Expression
	loadExpression()
}

type loadExpression struct {
	expression
}

func (node *loadExpression) loadExpression() {}

// LoadLocationExpression loads a location's address, producing a pointer that can be dereferenced.
type LoadLocationExpression struct {
	loadExpression
	Object *Expression `json:"object,omitempty"` // the `this` object, in the case of class properties.
	Name   *Identifier `json:"name"`             // the name of the member to load.
}

// LoadDynamicExpression dynamically loads either a variable or a function, by name, from an object.
type LoadDynamicExpression struct {
	loadExpression
	Object Expression `json:"object"` // the object from which to load the property.
	Name   Expression `json:"name"`   // the dynamically evaluated name of the property to load.
}

/* Functions */

type CallExpression interface {
	Expression
	GetArguments() *[]Expression // the list of arguments in sequential order.
}

type callExpression struct {
	expression
	Arguments *[]Expression `json:"arguments,omitempty"`
}

// NewExpression allocates a new object and calls its constructor.
type NewExpression struct {
	callExpression
	Type *Identifier // the object type to allocate.
}

// InvokeFunction invokes a target expression that must evaluate to a function.
type InvokeFunctionExpression struct {
	callExpression
	Function Expression `json:"function"` // a function to invoke (of a function type).
}

// LambdaExpression creates a lambda, a sort of "anonymous" function, that evaluates to a function type.
type LambdaExpression struct {
	expression
	Signature  symbols.TypeToken       `json:"signature"`  // the function signature type.
	Parameters []symbols.VariableToken `json:"parameters"` // the parameter variables.
	Body       *Block                  `json:"body"`       // the lambda's body block.
}

/* Operators */

// UnaryOperatorExpression is the usual C-like unary operator.
type UnaryOperatorExpression struct {
	expression
	Operator UnaryOperator `json:"operator"` // the operator type.
	Operand  Expression    `json:"operand"`  // the right hand side operand.
	Postfix  bool          `json:"postfix"`  // whether this is a postfix operator (only legal for UnaryPfixOperators).
}

// UnaryOperator is the full set of unary operator tokens.  Note that MuIL doesn't care about precedence.  The MetaMu
// compilers must present expression in the order in which they should be evaluated through an in-order AST tree walk.
type UnaryOperator string

const (
	// Prefix-only operators:
	OpDereference UnaryOperator = "*"
	OpAddressof                 = "&"
	OpUnaryPlus                 = "+"
	OpUnaryMinus                = "-"
	OpLogicalNot                = "!"
	OpBitwiseNot                = "~"

	// These are permitted to be prefix or postfix:
	OpPlusPlus   = "++"
	OpMinusMinus = "--"
)

// BinaryOperatorExpression is the usual C-like binary operator (assignment, logical, operator, or relational).
type BinaryOperatorExpression struct {
	expression
	Left     Expression     `json:"left"`     // the left hand side.
	Operator BinaryOperator `json:"operator"` // the operator.
	Right    Expression     `json:"right"`    // the right hand side.
}

// All of the available arithmetic operators.
type BinaryOperator string

const (
	// Arithmetic operators:
	OpAdd          BinaryOperator = "+"
	OpSubtract                    = "-"
	OpMultiply                    = "*"
	OpDivide                      = "/"
	OpRemainder                   = "%"
	OpExponentiate                = "**"

	// Assignment operators:
	OpAssign                  = "="
	OpAssignSum               = "+="
	OpAssignDifference        = "-="
	OpAssignProduct           = "*="
	OpAssignQuotient          = "/="
	OpAssignRemainder         = "%="
	OpAssignExponentiation    = "**="
	OpAssignBitwiseShiftLeft  = "<<="
	OpAssignBitwiseShiftRight = ">>="
	OpAssignBitwiseAnd        = "&="
	OpAssignBitwiseOr         = "|="
	OpAssignBitwiseXor        = "^="

	// Bitwise operators:
	OpBitwiseShiftLeft  = "<<"
	OpBitwiseShiftRight = ">>"
	OpBitwiseAnd        = "&"
	OpBitwiseOr         = "|"
	OpBitwiseXor        = "^"

	// Conditional operators:
	OpLogicalAnd = "&&"
	OpLogicalOr  = "||"

	// Relational operators:
	OpLt        = "<"
	OpLtEquals  = "<="
	OpGt        = ">"
	OpGtEquals  = ">="
	OpEquals    = "=="
	OpNotEquals = "!="
)

/* Type Testing */

// CastExpression handles both nominal and structural casts, and will throw an exception upon failure.
type CastExpression struct {
	expression
	Expression Expression        `json:"expression"` // the source expression.
	Type       symbols.TypeToken `json:"type"`       // the target type.
}

// IsInstExpression checks an expression for compatibility with the given type token, and evaluates to a bool.
type IsInstExpression struct {
	expression
	Expression Expression        `json:"expression"` // the source expression.
	Type       symbols.TypeToken `json:"type"`       // the target type.
}

// TypeOfExpression gets the type token -- just a string -- of a particular type at runtime.
type TypeOfExpression struct {
	expression
	Expression Expression `json:"expression"` // the source expression
}

/* Miscellaneous */

// ConditionalExpression evaluates to either a consequent or alternate based on a predicate condition.
type ConditionalExpression struct {
	expression
	Condition  Expression `json:"condition"`  // a `bool` conditional expression.
	Consequent Expression `json:"consequent"` // the expression to evaluate to if `true`.
	Alternate  Expression `json:"alternate"`  // the expression to evaluate to if `false`.
}

// SequenceExpression allows composition of multiple expressions into one.  It evaluates to the last one.
type SequenceExpression struct {
	expression
	Expressions []Expression `json:"expressions"`
}
